go/store/datas/pull/puller.go

```diff
@@ -296,139 +300,82 @@ func (p *Puller) Pull(ctx context.Context) error {
-			wrErr := p.wr.Close()
-			if err == nil {
-				err = wrErr
-			}
-		}()
+	const batchSize = 64 * 1024
+	tracker := NewPullChunkTracker(ctx, p.hashes, TrackerConfig{
+		BatchSize: batchSize,
+		HasManyer: p.sinkDBCS,
+	})
 
+	// One thread calls ChunkFetcher.Get on each batch.
+	eg.Go(func() error {
 		for {
 			toFetch, hasMore, err := tracker.GetChunksToFetch()
 			if err != nil {
 				return err
 			}
 			if !hasMore {
-				break
+				return p.rd.CloseSend()
 			}
 
-			err = p.getCmp(ctx, toFetch, tracker)
+			atomic.AddUint64(&p.stats.totalSourceChunks, uint64(len(toFetch)))
+			err = p.rd.Get(ctx, toFetch)
 			if err != nil {
 				return err
 			}
 		}
-
-		return nil
 	})
 
-	return eg.Wait()
-}
-
-// batchNovel returns a slice of |batch| size HashSets and partial |remainder| HashSet.
-func batchNovel(absent hash.HashSet, batch int) (remainder hash.HashSet, batches []hash.HashSet) {
-	curr := make(hash.HashSet, batch)
-	for h := range absent {
-		curr.Insert(h)
-		if curr.Size() >= batch {
-			batches = append(batches, curr)
-			curr = make(hash.HashSet, batch)
-		}
-	}
-	remainder = curr
-	return
-}
-
-func (p *Puller) getCmp(ctx context.Context, batch hash.HashSet, tracker *PullChunkTracker) error {
-	found := make(chan nbs.CompressedChunk, 4096)
-	processed := make(chan nbs.CompressedChunk, 4096)
-
-	atomic.AddUint64(&p.stats.totalSourceChunks, uint64(len(batch)))
-	eg, ctx := errgroup.WithContext(ctx)
+	// One thread reads the received chunks, walks their addresses and writes them to the table file.
 	eg.Go(func() error {
-		err := p.srcChunkStore.GetManyCompressed(ctx, batch, func(ctx context.Context, c nbs.CompressedChunk) {
-			atomic.AddUint64(&p.stats.fetchedSourceBytes, uint64(len(c.FullCompressedChunk)))
-			atomic.AddUint64(&p.stats.fetchedSourceChunks, uint64(1))
-			select {
-			case found <- c:
-			case <-ctx.Done():
-			}
-		})
-		if err != nil {
-			return err
-		}
-		close(found)
-		return nil
-	})
-
-	eg.Go(func() error {
-	LOOP:
 		for {
-			select {
-			case cmpChnk, ok := <-found:
-				if !ok {
-					break LOOP
-				}
-
-				chnk, err := cmpChnk.ToChunk()
-				if err != nil {
-					return err
-				}
-				err = p.waf(chnk, func(h hash.Hash, _ bool) error {
-					tracker.Seen(h)
-					return nil
-				})
-				if err != nil {
-					return err
-				}
-				select {
-				case processed <- cmpChnk:
-				case <-ctx.Done():
-					return ctx.Err()
-				}
-			case <-ctx.Done():
-				return ctx.Err()
+			cChk, err := p.rd.Recv(ctx)
+			if err == io.EOF {
+				// This means the requesting thread
+				// successfully saw all chunk addresses and
+				// called CloseSend and that all requested
+				// chunks were successfully delivered to this
+				// thread. Calling wr.Close() here will block
+				// on uploading any table files and will write
+				// the new table files to the destination's
+				// manifest.
+				return p.wr.Close()
+			}
+			if err != nil {
+				return err
+			}
+			if len(cChk.FullCompressedChunk) == 0 {
```

Comment:
The only way this would happen would be if a CompressedChunk{} was returned, which should always be accompanied by an error. Is this out of an abundance of caution, or is there another case where this will come up?